### ISSUE TITLE
PLIP: Reuse code between Portlets, Mosaic Tiles, Views and Cover Tiles

### ISSUE THREAD

**Issue by rodfersou** [plone/Products.CMFPlone#1755]
## Abstract
Move small applets in different layout machinery existent in Plone

## Motivation
Some applets like portlet Calendar are really complex, and there is no point to rewrite this code when need it in other places of Plone

## Assumptions
Define where the code should be centralized, and what is the better way to share code and template.

## Proposal & Implementation
Need suggestions.

## Deliverables
One place for code and templates, reused for Portlets, Mosaic Tiles, Views, and Cover Tiles in a easy way.

## Risks
Depend on how we are going to achieve this.

## Participants
@rodfersou @hvelarde

We started with an implementation using mixins and macros, but it was pointed that this is not the better way to go https://github.com/plone/plone.app.event/pull/243

**Comment by hvelarde** [comment]
we wanted to do that during the Mephisto Sprint but obviously that was not a good idea as developers were focused on other things.

@datakurre sugested to move the resulting code to [plone.app.layout](https://github.com/plone/plone.app.layout); @thet and @jensens didn't like the idea of mixins and suggested adapters as an option but we have no experience on that and need some code examples.

**Comment by jensens** [comment]
So we need a snippet for that which is code+template, has access to the context and request... 

So just a view, right? 

Then the view can be rendered in different scenarios such as portlet, tile, viewlet and so on by just looking it up and calling it in the template maybe like so:

```
<div tal:replace="structure: context/@@plone.snippets.calendar" />
```

This could be looked up and called in code too.

Just an idea, open for better/simpler solutions.

**Comment by jensens** [comment]
Since parameter passing is needed it could be turned into traversal:

`<div tal:replace="structure: context/@@plone.snippets.calendar/hash/1234567890/otherkey/othervalue" />
`

Also just an idea. This would work with parametrized JS as well

**Comment by thet** [comment]
I've done that like described by @jensens numerous times, except for passing parameters. But for parameters, you can also do:
```
<tal:init define="calendar_view nocall:context/@@plone.snippets.calendar">
    <div tal:replace="structure python:calendar_view.__call__(context, parameter_1)"/>
    <div tal:define="items python:calendar_view.get_items(parameter_2='something')">
        <div tal:repeat="item items" tal:content="item/title"/>
    </div>
</tal:init>
```

just dummy code and also just an idea...

**Comment by jensens** [comment]
> ```<div tal:replace="structure python:calendar_view.__call__(context, parameter_1)"/>```

I dont think this works like so. Anyway, the overall goal is to have a generic view-like thing which is in fact just a callable.

**Comment by rodfersou** [comment]
fine, I got the idea, thank you, and where is the right place to put then? I mean.. a place where in the future could be where to put every portlet / tile views

**Comment by gforcada** [comment]
@thet no ``python:`` in templates pretty please :-) that has a high performance impact

**Comment by jensens** [comment]
@gforcada, do you have measurements here? In restricted templates that may be true (i never profiled this with chameleon), but with chameleon and unrestricted templates (which we are talking about) from my experience tales expressions are slower than native python. This is true in pyramid as well btw, where python expressions are default.

**Comment by gforcada** [comment]
@jensens what are restricted/unrestricted templates? AFAIK any template has to go through the template engine (zope.tal, chameleon) and they are patched to go trough RestrictedPython, right? That's what I read about it, and since then moved as much logic as possible to python code(views) rather than on the templates

**Comment by jensens** [comment]
@gforcada From my understanding and from a look at ``five.pt``: No, they are not all restricted Python. 

If you look at ``five.pt`` where all the magic happens, it makes a difference if your code lives in the zodb and is untrusted (zope2), or if it is trusted code on the file-system via zcml (called zope3).

https://github.com/zopefoundation/five.pt/blob/master/src/five/pt/engine.py#L57-L77
and
https://github.com/zopefoundation/five.pt/blob/master/src/five/pt/engine.py#L121-L123

This also explains why templates customized through the ZMI ``portal_view_customizations`` do often stop working and start raising unauthorized if they used untrusted Python features before.

So, in short 
- filesystem-templates = fast = python expressions are good, native python parsing, no extra parsing needed.
- zodb templates = slow, it should not make a big difference compare to TALES, but I have no numbers here.

**Comment by rodfersou** [comment]
@jensens interesting

If I understand well the right place to create the calendar view is at plone.app.events package, and where can I put the tile rss code as a view?

**Comment by jensens** [comment]
Closing as outdated (10 years old). Portlets are deprecated in Plone 6 and tiles/Mosaic never became core. Architecture moved past this proposal.
If you think this issue is still relevant please reopen it. Thank you for the original report!
